src/v/features/feature_migrator.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "feature_migrator.h"
+
+#include "cluster/controller.h"
+#include "cluster/feature_manager.h"
+#include "features/feature_table.h"
+#include "features/logger.h"
+#include "ssx/future-util.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <seastar/core/sleep.hh>
+
+#include <type_traits>
+
+using namespace std::chrono_literals;
+
+namespace features {
+
+ss::future<> feature_migrator::stop() {
+    if (_as) {
+        ss::abort_source& as = *_as;
+        vassert(as.abort_requested(), "Stopped without requesting abort");
+    }
+    return _gate.close();
+}
+
+void feature_migrator::start(ss::abort_source& as) {
+    _as = as;
+    features::feature_table& ft = _controller.get_feature_table().local();
+    if (ft.is_active(get_feature())) {
+        vlog(
+          featureslog.trace,
+          "Feature {} already active, no migration required.",
+          get_feature());
+        return;
+    }
+
+    ssx::spawn_with_gate(_gate, [this]() { return do_migrate(); });
+}
+
+ss::future<> feature_migrator::do_migrate() {
+    features::feature_table& ft = _controller.get_feature_table().local();
+    if (ft.is_active(get_feature())) {
+        vlog(
+          featureslog.trace,
+          "Feature {} already active, no migration required.",
+          get_feature());
+        co_return;
+    }
+
+    co_await ft.await_feature_preparing(get_feature(), *_as);
```

Comment:
The assumption is that start() has already been called: this is actually jankier than necessary, an earlier version required start() to take the abort source (not constructor) because the old groups migrator worked that way, but the way the code ended up, we could just pass the abort source in the constructor and de-optionalize it